### PR TITLE
[docs] point in time restore

### LINKED ISSUE #7122: PITR: Docs on how to use PITR, history retention, snapshot intervals
After we have #7121, we should play around with it and prepare some BETA docs.

### PR DESCRIPTION
**Pull request by polarweasel**
Fixes #7122

Deploy preview: https://deploy-preview-7912--infallible-bardeen-164bc9.netlify.app/latest/manage/backup-restore/point-in-time-restore-ysql/

### CODE REVIEW

File: docs/content/latest/manage/backup-restore/point-in-time-restore-ysql.md [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+
+    ```sh
+    $ bin/yb-admin list_snapshots
+    ```
+
+    ```output
+    Snapshot UUID                         State
+    bb5fc435-a2b9-4f3a-a510-0bacc6aebccf  COMPLETE
+    No snapshot restorations
+    ```
```
Comment (bmatican) [inline docs/content/latest/manage/backup-restore/point-in-time-restore-ysql.md:240]: not sure if intended, but output and shell commands look the same to me, except for the COPY button...would be nice if in our docs infra, they were made to be visibly different somehow

File: docs/content/latest/manage/backup-restore/point-in-time-restore-ysql.md [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+
+1. Restore the latest snapshot to the timestamp you obtained before you deleted the data:
+
+    ```sh
+    # NOTE add a zero (0) to the end of the timestamp!
```
Comment (bmatican) [inline docs/content/latest/manage/backup-restore/point-in-time-restore-ysql.md:245]: this would be great to rework a bit to take the microsecond with 5 digit precision output and instead multiply by 1000000 here

File: docs/content/latest/manage/backup-restore/point-in-time-restore-ysql.md [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+    $ bin/yb-admin restore_snapshot bb5fc435-a2b9-4f3a-a510-0bacc6aebccf 1617670679185100
+
+1. Verify the snapshot is restored:
```
Comment (bmatican) [inline docs/content/latest/manage/backup-restore/point-in-time-restore-ysql.md:248]: maybe say / add "in RESTORED state"

File: docs/content/latest/manage/backup-restore/point-in-time-restore-ysql.md [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+    ```
+
+    ```output
+    Snapshot UUID                     State
+    bb5fc435-a2b9-4f3a-a510-0bacc6aebccf  COMPLETE
+    Restoration UUID                      State
+    bd7e4e52-b763-4b95-87ce-9399e1ac206e  RESTORED
+    ```
+
+1. Verify the data is restored:
```
Comment (bmatican) [inline docs/content/latest/manage/backup-restore/point-in-time-restore-ysql.md:261]: maybe a comment calling out how the bad row is not present?

File: docs/content/latest/manage/backup-restore/point-in-time-restore-ysql.md [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+            1221 | John Smith     | Marketing  |  50000
+            1222 | Bette Davis    | Sales      |  55000
+    (4 rows)
+    ```
+
+### Restore from a relative time
+
+In addition to restoring to a particular timestamp, you can also restore to a relative time, such as "ten minutes ago". In this example, you'll delete some data from the existing `employees` table, then restore the state of the database to what it was five minutes prior.
+
+1. Wait five minutes after you complete the steps in the previous section. This is so that you can use a known relative time for the restore.
```
Comment (bmatican) [inline docs/content/latest/manage/backup-restore/point-in-time-restore-ysql.md:281]: not sure if worth mentioning instead / as well, the side effect, that if you have not waited 5 minutes, you might restore to the state of the database, before you even inserted the initial rows?

File: docs/content/latest/manage/backup-restore/point-in-time-restore-ysql.md [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+            1224 | John Zimmerman | Sales      |  60000
+            1221 | John Smith     | Marketing  |  50000
+            1222 | Bette Davis    | Sales      |  55000
+    (4 rows)
+    ```
+
+1. Restore the snapshot you created earlier:
+
+    ```sh
+    $ bin/yb-admin restore_snapshot bb5fc435-a2b9-4f3a-a510-0bacc6aebccf "minus 3m"
```
Comment (bmatican) [inline docs/content/latest/manage/backup-restore/point-in-time-restore-ysql.md:303]: nit: you say 5min above, but wait 3 minutes here

cc @nspiegelberg is this the correct new syntax btw? I recall you saying on Slack that you need not quote minus?

File: docs/content/latest/manage/backup-restore/point-in-time-restore-ysql.md [hunk 7]
```diff
@@ -0,0 +1,331 @@
...
+## Limitations
+
+This is a BETA feature, and is in active development. Currently, you can **restore data only**. The feature doesn't support metadata; in other words, it won't currently roll back operations such as CREATE, ALTER, TRUNCATE, and DROP TABLE.
+
+Development for this feature is tracked in [issue 7120](https://github.com/yugabyte/yugabyte-db/issues/7120). Some forthcoming features include:
+
+* Automatic configuration
+* Early metadata support, such as undoing CREATE TABLE operations
+* More complete metadata support: TRUNCATE, DROP, ALTER
+* Per-database restore for YSQL
```
Comment (bmatican) [inline docs/content/latest/manage/backup-restore/point-in-time-restore-ysql.md:331]: not sure if to list these out like this, since it seems weird to read "early support for X" vs "more support for Y"

if you want to call something out, maybe split into
- support for undoing metadata operations, such as CREATE, ALTER, TRUNCATE or DROP TABLE
- schedules to ensure in-cluster snapshots are taken at user defined intervals
- options to restore at different granularities, such as a complete YSQL database vs the whole YCQL dataset

**Review by bmatican — CHANGES_REQUESTED** [review]

File: docs/content/latest/manage/backup-restore/point-in-time-restore-ysql.md (hunk 1 above)
Comment (polarweasel) [reply docs/content/latest/manage/backup-restore/point-in-time-restore-ysql.md:240]: Nope... source highlighting for shell is pretty limited

File: docs/content/latest/manage/backup-restore/point-in-time-restore-ysql.md [hunk 8]
```diff
@@ -0,0 +1,331 @@
...
+
+Point-in-time recovery (also referred to here as PITR) and incremental backups go hand in hand. These two features help in recovering from a number of error or failure scenarios by allowing the database to be rolled back to a specific point in time (in the past). The rollback is done from the last full backup, with the updates since that full backup being replayed till the desired point in time.
+
+Point-in-time restores and incremental backups depend on full backups (also referred to as base backups). A full backup, as the name suggests, is a complete transactional backup of data up to a certain point in time. The entire data set in the database is backed up for the set of namespaces/tables chosen by the user. Read more about the design for a full, distributed backup. Full backups are deemed expensive for the following reasons:
+
+The performance of the database could be adversely affected. There may be a high latency of foreground operations, decreased throughput when the backup is happening. The data set size may be large, requiring more time, bandwidth and disk space to perform the backup.
+
+## Recovery scenarios
```
[the remark this answers is not in the document]
Comment (polarweasel) [reply docs/content/latest/manage/backup-restore/point-in-time-restore-ysql.md:39]: This isn't for you, it's for OUTSIDE readers. So yes, I copied some from the design doc. I'll slim it down.

File: docs/content/latest/manage/backup-restore/point-in-time-restore-ysql.md [hunk 9]
```diff
@@ -0,0 +1,331 @@
...
+    bb5fc435-a2b9-4f3a-a510-0bacc6aebccf  COMPLETE
+    No snapshot restorations
+    ```
+
+### Restore from an absolute time
+
+1. Get a timestamp:
+
+    ```sql
+    yugabyte=# select extract(epoch from now())*100000 as current_timestamp;
```
[the remark this answers is not in the document]
Comment (polarweasel) [reply docs/content/latest/manage/backup-restore/point-in-time-restore-ysql.md:199]: Added a note, removed the multiplication...

File: docs/content/latest/manage/backup-restore/point-in-time-restore-ysql.md (hunk 6 above)
Comment (polarweasel) [reply docs/content/latest/manage/backup-restore/point-in-time-restore-ysql.md:303]: I did that deliberately for a bit of a buffer, but you're right it's a bit odd. Changed to 5m.

File: docs/content/latest/manage/backup-restore/point-in-time-restore-ysql.md (hunk 7 above)
Comment (polarweasel) [reply docs/content/latest/manage/backup-restore/point-in-time-restore-ysql.md:331]: Updated...

File: docs/content/latest/manage/backup-restore/point-in-time-restore-ysql.md (hunk 2 above)
Comment (polarweasel) [reply docs/content/latest/manage/backup-restore/point-in-time-restore-ysql.md:245]: Done

File: docs/content/latest/manage/backup-restore/point-in-time-restore-ysql.md (hunk 3 above)
Comment (polarweasel) [reply docs/content/latest/manage/backup-restore/point-in-time-restore-ysql.md:248]: Done

File: docs/content/latest/manage/backup-restore/point-in-time-restore-ysql.md (hunk 4 above)
Comment (polarweasel) [reply docs/content/latest/manage/backup-restore/point-in-time-restore-ysql.md:261]: Done

File: docs/content/latest/manage/backup-restore/point-in-time-restore-ysql.md (hunk 5 above)
Comment (polarweasel) [reply docs/content/latest/manage/backup-restore/point-in-time-restore-ysql.md:281]: I'll adjust pending your/Nicolas's response in slack :)

Comment (polarweasel) [reply docs/content/latest/manage/backup-restore/point-in-time-restore-ysql.md]: Updated

File: docs/content/latest/manage/backup-restore/point-in-time-restore-ycql.md [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+    ```output
+    Snapshot UUID                         State
+    bb5fc435-a2b9-4f3a-a510-0bacc6aebccf  COMPLETE
+    No snapshot restorations
+    ```
+
+1. Restore the latest snapshot to the timestamp you obtained before you deleted the data, at a terminal prompt:
+
+    ```sh
+    $ bin/yb-admin restore_snapshot bb5fc435-a2b9-4f3a-a510-0bacc6aebccf 1617670679185100
```
Comment (nspiegelberg) [inline docs/content/latest/manage/backup-restore/point-in-time-restore-ycql.md:158]: The timestamp '1617670679185100' doesn't seem to have a source.  Maybe you should use that as the example output somewhere in the 'Get a timestamp' section.

File: docs/content/latest/manage/backup-restore/point-in-time-restore-ycql.md [hunk 11]
```diff
@@ -0,0 +1,268 @@
...
+            1222 |    Bette Davis |      Sales |  55000
+
+    (4 rows)
+    ```
+
+### Restore from a relative time
+
+In addition to restoring to a particular timestamp, you can also restore to a relative time, such as "ten minutes ago". In this example, you'll delete some data from the existing `employees` table, then restore the state of the database to what it was five minutes prior.
+
+When you specify a relative time, you can specify any or all of _days_, _hours_, and _minutes_. For example:
```
Comment (nspiegelberg) [inline docs/content/latest/manage/backup-restore/point-in-time-restore-ycql.md:194]: For entering in these times, we support:
1. ISO 8601: '3d 4h 5m 6s'
2. Abbreviated Postgres: '3 d 4 hrs 5 mins 6 secs'
3. Traditional Postgres: '3 days 4 hours 5 minutes 6 seconds'
4. SQL Standard: 'D H:M:S'

I'm not sure if you want to comment that here.

File: docs/content/latest/manage/backup-restore/point-in-time-restore-ycql.md [hunk 12]
```diff
@@ -0,0 +1,268 @@
...
+            1224 | John Zimmerman |      Sales |  60000
+            1221 |     John Smith |  Marketing |  50000
+            1222 |    Bette Davis |      Sales |  55000
+
+    (4 rows)
+    ```
+
+## Limitations
+
+This is a BETA feature, and is in active development. Currently, you can **restore data only**. The feature doesn't support metadata; in other words, it won't currently roll back operations such as CREATE, ALTER, TRUNCATE, and DROP TABLE.
```
Comment (nspiegelberg) [inline docs/content/latest/manage/backup-restore/point-in-time-restore-ycql.md:261]: Do we want to say "metadata / schema changes"

Comment (nspiegelberg) [reply docs/content/latest/manage/backup-restore/point-in-time-restore-ycql.md:261]: Maybe reword to "in other words, you currently shouldn't roll back beyond operations ... because it is unsupported."

File: docs/content/latest/manage/backup-restore/point-in-time-restore-ycql.md (hunk 10 above)
Comment (polarweasel) [reply docs/content/latest/manage/backup-restore/point-in-time-restore-ycql.md:158]: Added output samples for the various shell options, and grabbed one for this example. Thanks!

File: docs/content/latest/manage/backup-restore/point-in-time-restore-ycql.md (hunk 11 above)
Comment (polarweasel) [reply docs/content/latest/manage/backup-restore/point-in-time-restore-ycql.md:194]: Excellent! Yes, added.

File: docs/content/latest/manage/backup-restore/point-in-time-restore-ycql.md (hunk 12 above)
Comment (polarweasel) [reply docs/content/latest/manage/backup-restore/point-in-time-restore-ycql.md:261]: Updated!

File: docs/content/latest/manage/backup-restore/point-in-time-restore-ycql.md [hunk 13]
```diff
@@ -0,0 +1,288 @@
...
+    $ bin/yb-admin list_snapshots
+    ```
+
+    ```output
+    Snapshot UUID                         State
+    bb5fc435-a2b9-4f3a-a510-0bacc6aebccf  COMPLETE
+    No snapshot restorations
+    ```
+
+### Restore from an absolute time
```
Comment (nspiegelberg) [inline docs/content/latest/manage/backup-restore/point-in-time-restore-ycql.md:105]: For the absolute time, we should also add a link to https://docs.yugabyte.com/latest/api/ycql/type_datetime/#timestamp , because the restore_timestamp also takes that string input.

**Review by nspiegelberg — CHANGES_REQUESTED** [review]

**Review by nspiegelberg — APPROVED** [review]

**Review by nspiegelberg — APPROVED** [review]

File: docs/content/latest/manage/backup-restore/point-in-time-restore-ycql.md (hunk 13 above)
Comment (polarweasel) [reply docs/content/latest/manage/backup-restore/point-in-time-restore-ycql.md:105]: (Done)
